### PR TITLE
Fix CDM timer issues

### PR DESCRIPTION
**Pull request by glennguy**
## Description

Provide proper solution to the CDM timer calls.

## Motivation and context

For a long time now there have been crashes related to the CDM timer finish callback being executed in a separate thread after CdmAdapter object (and even inputstream.adaptive) has been destroyed.

The current code attempts to wait in the CdmAdapter destructor for the timers to finish, however the destructor isn't even called because of the existing shared_ptr references in the timer threads. wv_decrypter just removes its shared_ptr reference to CdmAdapter and continues destructing regardless.

The new code interrupts any timers that are waiting and removes the references when closing the session. Streams seem noticeably quicker to stop playback.

Additionally thanks to use of std::async to perform the timer task rather than std::thread detach we're not creating a new thread every second. std::async implementation uses its own thread pool of sorts. 

## How has this been tested?

Tested on Kayo add-on by starting stopping multiple streams, and also confirming playback continues after the 2 minute mark - if timer expired not called, the license renewal will fail and stream will stop decrypting at this point.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the Wiki documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by tmm1** [comment]
Nice! I've noticed both the crashes and slow stops. Will test this out.

**Comment by glennguy** [comment]
I've had a bunch of kayo sports users test with success, I hope this part of the codebase will be good now!

**Comment by tmm1** [comment]
LGTM in my testing

File: wvdecrypter/cdm/media/cdm/cdm_adapter.cc [hunk 1]
```diff
@@ -70,19 +70,17 @@ void* GetCdmHost(int host_interface_version, void* user_data)
...
-  timer_thread_running  = true;
-  uint64_t waited = 0;
-  std::this_thread::sleep_for(std::chrono::milliseconds(delay));
-  if (!exit_thread_flag) 
   {
-    adp->TimerExpired(context);
+    std::unique_lock<std::mutex> lock(m_closeSessionMutex);
+    auto predicate = [this] { return ms_isCloseSession.load(); };
+    const bool isClosing =
+      ms_sessionClosingCond.wait_for(lock, std::chrono::milliseconds(delay), predicate);
```
Comment (CastagnaIT) [inline wvdecrypter/cdm/media/cdm/cdm_adapter.cc]: predicate can stay inline
```suggestion
    const bool isClosing = ms_sessionClosingCond.wait_for(
        lock, std::chrono::milliseconds(delay), [this] { return ms_isCloseSession.load(); });
```

File: wvdecrypter/cdm/media/cdm/cdm_adapter.cc [hunk 2]
```diff
@@ -333,17 +325,19 @@ void CdmAdapter::CloseSession(uint32_t promise_id,
   const char* session_id,
   uint32_t session_id_size)
 {
-  exit_thread_flag = true;
-  while (timer_thread_running) 
   {
-    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    std::lock_guard<std::mutex> lock(m_closeSessionMutex);
+    CdmAdapter::ms_isCloseSession = true;
```
Comment (CastagnaIT) [inline wvdecrypter/cdm/media/cdm/cdm_adapter.cc]: there is a reason why you add class name prefix?

File: wvdecrypter/cdm/media/cdm/cdm_adapter.cc [hunk 3]
```diff
@@ -507,7 +501,21 @@ cdm::Buffer* CdmAdapter::Allocate(uint32_t capacity)
...
+  {
+    std::lock_guard<std::mutex> lock(m_closeSessionMutex);
+    for (auto i = 0; i < m_futures.size(); i++) {
+      if (!m_futures[i].valid() || m_futures[i].wait_for(std::chrono::milliseconds(0)) ==
+        std::future_status::ready)
+      {
+        m_futures.erase(m_futures.begin() + i);
+        i--;
+      }
+    }
```
Comment (CastagnaIT) [inline wvdecrypter/cdm/media/cdm/cdm_adapter.cc]: why not use iterator?
```suggestion
    for (auto& itFuture = m_futures.begin(); itFuture != m_futures.end();)
    {
      if (!itFuture->valid() ||
          itFuture->wait_for(std::chrono::milliseconds(0)) == std::future_status::ready)
      {
        itFuture = m_futures.erase(itFuture);
      }
      else
        itFuture++;
    }
```

File: wvdecrypter/cdm/media/cdm/cdm_adapter.cc [hunk 4]
```diff
@@ -507,7 +501,21 @@ cdm::Buffer* CdmAdapter::Allocate(uint32_t capacity)
...
+  }
+  auto future = std::async(std::launch::async, &CdmAdapter::timerfunc, shared_from_this(), this,
```
Comment (CastagnaIT) [inline wvdecrypter/cdm/media/cdm/cdm_adapter.cc]:
```suggestion
  std::future<void> future = std::async(std::launch::async, &CdmAdapter::timerfunc, shared_from_this(), this,
```

File: wvdecrypter/cdm/media/cdm/cdm_adapter.h [hunk 5]
```diff
@@ -235,7 +237,12 @@ class CdmAdapter : public std::enable_shared_from_this<CdmAdapter>
   std::string cdm_path_;
   std::string cdm_base_path_;
   CdmAdapterClient *client_;
-  std::mutex client_mutex_, decrypt_mutex_;
+  std::mutex client_mutex_;
+  std::mutex decrypt_mutex_;
+  std::mutex m_closeSessionMutex;
+  std::atomic<bool> ms_isCloseSession;
```
Comment (CastagnaIT) [inline wvdecrypter/cdm/media/cdm/cdm_adapter.h]:
```suggestion
  std::atomic<bool> m_isClosingSession;
```

File: wvdecrypter/cdm/media/cdm/cdm_adapter.h [hunk 6]
```diff
@@ -235,7 +237,12 @@ class CdmAdapter : public std::enable_shared_from_this<CdmAdapter>
...
+  std::condition_variable ms_sessionClosingCond;
```
Comment (CastagnaIT) [inline wvdecrypter/cdm/media/cdm/cdm_adapter.h]:
```suggestion
  std::condition_variable m_sessionClosingCond;
```

File: wvdecrypter/cdm/media/cdm/cdm_adapter.h [hunk 7]
```diff
@@ -235,7 +237,12 @@ class CdmAdapter : public std::enable_shared_from_this<CdmAdapter>
...
+  std::vector<std::future<void>> m_futures;
```
Comment (CastagnaIT) [inline wvdecrypter/cdm/media/cdm/cdm_adapter.h]: variable name dont explain his use, could be an idea something like:
```suggestion
  std::vector<std::future<void>> m_asyncTimerTasks;
```

**Review by CastagnaIT — CHANGES_REQUESTED** [review]
some code style

**Comment by glennguy** [comment]
Thanks for feedback Stefano. Some of these like variable names are 'sausage factory' artifacts 😆

File: wvdecrypter/cdm/media/cdm/cdm_adapter.cc [hunk 8]
```diff
@@ -504,10 +485,38 @@ cdm::Buffer* CdmAdapter::Allocate(uint32_t capacity)
...
+}
+
 void CdmAdapter::SetTimer(int64_t delay_ms, void* context)
 {
   //LICENSERENEWAL
-  std::thread(timerfunc, shared_from_this(), delay_ms, context).detach();
+  // Clean up async tasks which have finished
+  {
+    std::lock_guard<std::mutex> lock(m_closeSessionMutex);
+    for (auto& itFuture = m_asyncTasks.begin(); itFuture != m_asyncTasks.end();)
```
Comment (CastagnaIT) [inline wvdecrypter/cdm/media/cdm/cdm_adapter.cc]: try check its possible auto& needs to be const
```suggestion
    for (const auto& itTimerTask = m_asyncTasks.begin(); itFuture != m_asyncTasks.end();)
```
maybe you can also rename
m_asyncTasks
to
m_asyncTimerTasks

**Comment by glennguy** [comment]
Ok changes made, tests are passing now too

File: wvdecrypter/cdm/media/cdm/cdm_adapter.cc [hunk 9]
```diff
@@ -504,10 +485,40 @@ cdm::Buffer* CdmAdapter::Allocate(uint32_t capacity)
...
+}
+
 void CdmAdapter::SetTimer(int64_t delay_ms, void* context)
 {
   //LICENSERENEWAL
-  std::thread(timerfunc, shared_from_this(), delay_ms, context).detach();
+  // Clean up async tasks which have finished
+  {
+    std::lock_guard<std::mutex> lock(m_closeSessionMutex);
+    LOG::Log(SSD::SSDDEBUG, "VEC: %u", m_asyncTimerTasks.size());
```
Comment (CastagnaIT) [inline wvdecrypter/cdm/media/cdm/cdm_adapter.cc]: you have forgot to remove this and the other one below

Comment (glennguy) [reply wvdecrypter/cdm/media/cdm/cdm_adapter.cc]: Sorry, fixed now. Need more sleep.

**Review by CastagnaIT — APPROVED** [review]
well done
